packages/gatsby-cli/src/reporter/redux/index.ts

```diff
@@ -46,6 +117,95 @@ export const dispatch = (action: ActionsUnion | Thunk): void => {
 
   store.dispatch(action)
 
+  // ignore diagnostic logs, otherwise diagnostic message itself will reset
+  // the timers
+  if (!displayingStuckStatusDiagnosticWarning) {
+    const currentStatus = store.getState().logs.status
+
+    // yuck, we have situation of circular dependencies here
+    // so this is `reporter` import is delayed until it's actually needed
+    const { reporter } = require(`../reporter`)
+
+    if (stuckStatusDiagnosticTimeoutDelay) {
+      if (stuckStatusDiagnosticTimer) {
+        clearTimeout(stuckStatusDiagnosticTimer)
+      }
+
+      if (displayedStuckStatusDiagnosticWarning) {
+        // using nextTick here to prevent infinite recursion (report.warn would
+        // result in another call of this function and so on)
+        process.nextTick(() => {
+          const activitiesDiagnosticsMessage = generateStuckStatusDiagnosticMessage()
+          reporter.warn(
+            `This is just diagnostic information (enabled by GATSBY_DIAGNOSTIC_STUCK_STATUS_TIMEOUT):\n\nThere was activity since last diagnostic message. Log action:\n\n${JSON.stringify(
+              action,
+              null,
+              2
+            )}\n\nCurrently Gatsby is in: "${
+              store.getState().logs.status
+            }" state.${
+              activitiesDiagnosticsMessage.length > 0
+                ? `\n\nActivities preventing Gatsby from transitioning to idle state:\n\n${activitiesDiagnosticsMessage}`
+                : ``
+            }`
+          )
+        })
+        displayedStuckStatusDiagnosticWarning = false
+      }
+
+      if (currentStatus === ActivityStatuses.InProgress) {
+        stuckStatusDiagnosticTimer = setTimeout(
+          function logStuckStatusDiagnostic() {
+            displayingStuckStatusDiagnosticWarning = true
+            reporter.warn(
+              `This is just diagnostic information (enabled by GATSBY_DIAGNOSTIC_STUCK_STATUS_TIMEOUT):\n\nGatsby is in "${
+                store.getState().logs.status
+              }" state without any updates for ${(
+                (stuckStatusDiagnosticTimeoutDelay ?? 0) / 1000
```

Comment:
If this was falsy (`null` or `0`) then it wouldn't pass the check on (new) line 129.
```suggestion
                stuckStatusDiagnosticTimeoutDelay / 1000
```

Reply:
This is to satisfy TS gods ;( But with your previous comment to use `0` and not `null` I will probably be able to drop this fallback)

Reply:
Done as part of https://github.com/gatsbyjs/gatsby/pull/26618/commits/e10a55644f3f0dd8a4bd419c7fa23075660e3e93